### ISSUE TITLE
Adding additional memory related detailed information 

### ISSUE THREAD

**Issue by EsQueEl-Fella** [trimble-oss/dba-dash#348]
This is a feature request.

Could we add additional memory related detailed information to the "Memory Clerks" Tab? Information such as the following useful sample queries mentioned in (https://learn.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/sys-dm-os-buffer-descriptors-transact-sql?view=sql-server-ver16)

SELECT COUNT(*)AS cached_pages_count  
    ,CASE database_id   
        WHEN 32767 THEN 'ResourceDb'   
        ELSE db_name(database_id)   
        END AS database_name  
FROM sys.dm_os_buffer_descriptors  
GROUP BY DB_NAME(database_id) ,database_id  
ORDER BY cached_pages_count DESC;  

and

SELECT COUNT(*)AS cached_pages_count   
    ,name ,index_id   
FROM sys.dm_os_buffer_descriptors AS bd   
    INNER JOIN   
    (  
        SELECT object_name(object_id) AS name   
            ,index_id ,allocation_unit_id  
        FROM sys.allocation_units AS au  
            INNER JOIN sys.partitions AS p   
                ON au.container_id = p.hobt_id   
                    AND (au.type = 1 OR au.type = 3)  
        UNION ALL  
        SELECT object_name(object_id) AS name     
            ,index_id, allocation_unit_id  
        FROM sys.allocation_units AS au  
            INNER JOIN sys.partitions AS p   
                ON au.container_id = p.partition_id   
                    AND au.type = 2  
    ) AS obj   
        ON bd.allocation_unit_id = obj.allocation_unit_id  
WHERE database_id = DB_ID()  
GROUP BY name, index_id   
ORDER BY cached_pages_count DESC;  

There also some interesting queries here (https://mssqlwiki.com/sqlwiki/sql-performance/troubleshooting-sql-server-memory/) which are of course a bit outdated as the SQL memory management is changed a bit but with some smaller adaptions we could update some of the useful ones and collect the data for reporting.

**Comment by DavidWiseman** [comment]
Hi,

Querying sys.dm_os_buffer_descriptors can be very expensive - especially on servers with large amounts of memory.  I came across an issue on one server where a monitoring tool was querying sys.dm_os_buffer_descriptors every 20 seconds and it was responsible for something like 90% of the CPU utilization on the server.
I have a server with 1TB of RAM - it takes over 1min to execute that first query.  The second would be more expensive and have to be executed in each DB.

It can be interesting to see what's in memory with sys.dm_os_buffer_descriptors but I would recommend running these queries ad hoc rather than capturing the data regularly.  

DBA Dash does have the flexibility to collect this data - if you really wanted to.  [Custom metrics are described here](https://github.com/trimble-oss/dba-dash/blob/main/Docs/OSPerformanceCounters.md).  Using that template you could create a stored proc like this in the DB that DBA Dash connects to:

```sql
CREATE OR ALTER PROC [dbo].[DBADash_CustomPerformanceCounters]
AS
-- Table variable isn't needed but it ensures the data is returned in the required format.  
DECLARE @Return AS TABLE(
	SnapshotDate DATETIME2(7) NOT NULL DEFAULT(SYSUTCDATETIME()),
	object_name NVARCHAR(128) NOT NULL,
	counter_name NVARCHAR(128) NOT NULL,
	instance_name NVARCHAR(128) NOT NULL,
	cntr_value DECIMAL(28,9) NOT NULL,
	cntr_type INT NOT NULL DEFAULT(65792)
)
-- ********************** Insert your own counters here **********************
-- Ommitting SnapshotDate and cntr_type columns as these are supplied by table defaults
INSERT INTO @Return
(
    object_name,
    counter_name,
    instance_name,
    cntr_value
)
SELECT 'os_buffer_descriptors',
	'Cached Pages Count',
	CASE database_id WHEN 32767 THEN 'ResourceDb' ELSE db_name(database_id) END AS database_name,
	COUNT(*) AS cached_pages_count
FROM sys.dm_os_buffer_descriptors
GROUP BY DB_NAME(database_id) ,database_id

-- ***************************************************************************

-- Return the data in the required format
SELECT SnapshotDate, -- Time in UTC
       object_name, -- Name of the object.  e.g. "User Counters"
       counter_name, -- Name of the counter.  e.g. "App Response Time (ms)"
       instance_name, -- Name of the instance.  e.g. "", "_Total", "DATABASE_NAME" etc
       cntr_value, -- The value of the couner (decimal)
       cntr_type -- follows the types ued in sys.dm_os_performanance_counters.  Use 65792 in most cases to use the value as is without further calculation.  Valid values: 65792,272696576,537003264,1073874176 
FROM @Return
```

The data would then be collected with the performance counters (every 1min by default) and available to view on the "Metrics" tab.

> **Warning**
> I would not recommend capturing this data due to the overhead it will add.

Hope this helps,

David

**Comment by EsQueEl-Fella** [comment]
I agree with the "Performance" related arguments. I am also aware of the time one needs to collect the data. On the other hand,

- Suggested queries were only a suggestion :D, maybe there are some queries out there which aren't so heavyduty. Those I suggested are only as a sample for a start.
- Running Adhoc means leaving the DBADash and do the job somewhere else. This eventually would mean the user must have many tools around which in return means in many cases chaos on the desktop.
- I thought of using the "Custom Metrics" before I write the Feature Request. The point here isn't how **one** user may adapt the application but how the application provides the generic useful features and the user can decide if he/she needs it or not. I probably will give it a try with "Custom Metrics" later on.
